File: gmcserver-server/src/main/java/me/vinceh121/gmcserver/entities/Device.java
/*
 * GMCServer, lightweight service to log, analyze and proxy Geiger counter data.
 * Copyright (C) 2020 Vincent Hyvert
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public for more details.
 *
 * You should have received a copy of the GNU Affero General Public
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package me.vinceh121.gmcserver.entities;

import java.util.Date;
import java.util.Map;

import org.bson.codecs.pojo.annotations.BsonIgnore;
import org.bson.types.ObjectId;

import com.mongodb.client.model.geojson.Point;

import io.vertx.core.json.JsonObject;

public class Device extends AbstractEntity {
	private String model, name, importedFrom;
	private Point location;
	private ObjectId owner, lastRecordId;
	private long gmcId;
	private boolean disabled;
	private Date lastEmailAlert = new Date(0L);
	private double stdDevAlertLimit = Double.NaN;
	private Map<String, Map<String, Object>> proxiesSettings;
	private Record lastRecord;

	public long getGmcId() {
		return this.gmcId;
	}

	public void setGmcId(final long gmcId) {
		this.gmcId = gmcId;
	}

	public ObjectId getOwner() {
		return this.owner;
	}

	public void setOwner(final ObjectId owner) {
		this.owner = owner;
	}

	public String getModel() {
		return this.model;
	}

	public void setModel(final String model) {
		this.model = model;
	}

	public String getName() {
		return this.name;
	}

	public void setName(final String name) {
		this.name = name;
	}

	public Point getLocation() {
		return this.location;
	}

	public void setLocation(final Point location) {
		this.location = location;
	}

	public boolean isDisabled() {
		return this.disabled;
	}

	public void setDisabled(final boolean disabled) {
		this.disabled = disabled;
	}

	public String getImportedFrom() {
		return this.importedFrom;
	}

	public void setImportedFrom(final String importedFrom) {
		this.importedFrom = importedFrom;
	}

	public Date getLastEmailAlert() {
		return this.lastEmailAlert;
	}

	public void setLastEmailAlert(final Date lastEmailAlert) {
		this.lastEmailAlert = lastEmailAlert;
	}

	public double getStdDevAlertLimit() {
		return this.stdDevAlertLimit;
	}

	public void setStdDevAlertLimit(final double stdDevAlertLimit) {
		this.stdDevAlertLimit = stdDevAlertLimit;
	}

	/**
	 * @return map where keys are proxy names (their simple class names, also keys
	 *         in GMCServer#getProxies) and values are proxy-specific settings
	 */
	public Map<String, Map<String, Object>> getProxiesSettings() {
		return this.proxiesSettings;
	}

	/**
	 * @return map where keys are proxy names (their simple class names, also keys
	 *         in GMCServer#getProxies) and values are proxy-specific settings
	 */
	public void setProxiesSettings(final Map<String, Map<String, Object>> proxiesSettings) {
		this.proxiesSettings = proxiesSettings;
	}

	public ObjectId getLastRecordId() {
		return lastRecordId;
	}

	public Device setLastRecordId(ObjectId lastRecordId) {
		this.lastRecordId = lastRecordId;
		return this;
	}

	/**
	 * Aggregated field
	 */
	@BsonIgnore
	public Record getLastRecord() {
		return lastRecord;
	}

	public Device setLastRecord(Record lastRecord) {
		this.lastRecord = lastRecord;
		return this;
	}

	@Override
	public JsonObject toJson() {
		final JsonObject obj = super.toJson();
		if (this.location != null) {
			obj.put("location", this.location.getCoordinates().getValues());
		}
		obj.put("owner", this.getOwner().toHexString());
		return obj;
	}

	@Override
	public JsonObject toPublicJson() {
		final JsonObject obj = super.toPublicJson();
		obj.remove("gmcId");
		obj.remove("proxiesSettings");
		obj.remove("lastEmailAlert");
		obj.remove("stdDevAlertLimit");
		obj.remove("lastRecordId");
		if (this.getLastRecord() == null) {
			obj.remove("lastRecord");
		}
		return obj;
	}

	public JsonObject toMapJson() {
		final JsonObject obj = new JsonObject();
		obj.put("id", this.getId().toHexString());
		obj.put("name", this.getName());
		obj.put("location", this.location.getCoordinates().getValues());
		obj.put("lastRecord", this.getLastRecord().toPublicJson());
		return obj;
	}
}

File: gmcserver-integration-tests/Device.js
/*
 * GMCServer, lightweight service to log, analyze and proxy Geiger counter data.
 * Copyright (C) 2020 Vincent Hyvert
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public for more details.
 *
 * You should have received a copy of the GNU Affero General Public
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

const fetch = require("node-fetch");
const { URL, LOG_URL, OBJECTID_REGEX } = require("./globals");
const { login, loginDevice } = require("./gmcapi");

const DEVICE_NAME = "My Test Device";
const DEVICE_POS = [51.1594650, -1.4439722];
const DEVICE_MODEL = "Node Fetch";

let device = undefined;

const createDevice = async () => {
	const expectedDevice = {
		name: DEVICE_NAME,
		model: DEVICE_MODEL,
		location: DEVICE_POS,
		owner: expect.stringMatching(OBJECTID_REGEX)
	};

	const token = await login();
	const res = await fetch(URL + "/device",
		{
			method: "POST",
			body: JSON.stringify({ name: DEVICE_NAME, position: DEVICE_POS, model: DEVICE_MODEL }),
			headers: {
				Authorization: token,
				"Content-Type": "application/json"
			}
		}
	);
	expect(res.status).toBe(200);

	const data = await res.json();
	device = data;
	expect(data).toMatchObject(expectedDevice);
};

const emptyTimeline = async () => {
	const token = await login();
	const res = await fetch(URL + "/device/" + device.id + "/timeline")
	expect(res.status).toBe(200);

	const data = await res.json();
	expect(data).toStrictEqual([]);
};

const logGmc = async () => {
	const DATE = new Date().getTime();
	const CPM = 42.0;
	const USV = 0.69;
	const ACPM = 42.69;

	const { deviceGmcId, userGmcId } = await loginDevice(device.id);

	const logRes = await fetch(LOG_URL + `/log2.asp?AID=${userGmcId}&GID=${deviceGmcId}&CPM=${CPM}&UsV=${USV}&aCpM=${ACPM}`);
	const logData = await logRes.text();
	expect(logData).toBe("OK.ERR0");

	const tlRes = await fetch(URL + "/device/" + device.id + "/timeline");
	const tlData = await tlRes.json();
	const rec = tlData[0];

	expect(rec).toMatchObject({ cpm: CPM, usv: USV, acpm: ACPM });
	expect(rec.date >= DATE).toBe(true);
};

const logSafecast = async () => {
	const DATE = new Date();
	const CPM = 42.0;
	const LON = 1.444;
	const LAT = 43.6045;

	const { deviceGmcId, userGmcId } = await loginDevice(device.id);

	const logRes = await fetch(LOG_URL + "/measurements.json?api_key=" + userGmcId, {
		method: "POST",
		body: JSON.stringify({
			device_id: deviceGmcId,
			unit: "cpm",
			value: CPM,
			captured_at: DATE,
			longitude: LON,
			latitude: LAT
		}),
		headers: { "Content-Type": "application/json" }
	});
	const logData = await logRes.text();
	expect(logRes.status).toBe(200);

	const tlRes = await fetch(URL + "/device/" + device.id + "/timeline");
	const tlData = await tlRes.json();
	const rec = tlData[0];

	expect(rec).toMatchObject({ cpm: CPM, location: [LON, LAT] });
	expect(rec.date >= DATE).toBe(true);
};

module.exports = [createDevice, emptyTimeline, logGmc, logSafecast];
